falkordb-cluster/cluster-entrypoint.sh

```diff
@@ -70,11 +70,120 @@ DATE_NOW=$(date +"%Y%m%d%H%M%S")
+  # Now the nodes should communitcate as expected.
+
+  if [[ -f "$DATA_DIR/nodes.conf" && -s "$DATA_DIR/nodes.conf" ]];then
+    discrepancy=0
+    while IFS= read -r line;do
+      if [[ $line =~ .*@0.* || $line =~ .*fail.* ]];then
+        discrepancy=$((discrepancy + 1))
+        hostname=$(echo "$line" | awk '{print $2}' | cut -d',' -f2| cut -d':' -f1)
+
+        tout=$(( $(date +%s) + 300 ))
+        while true;do
+          if [[ $(date +%s) -gt $tout ]];then 
+            echo "Timedout after 5 minutes while trying to ping $ip"
+            exit 1
+          fi
+
+          sleep 3
+
+          ip=$(getent hosts "$hostname" | awk '{print $1}')
+          PONG=$(redis-cli -h $(echo $hostname | cut -d'.' -f1) $AUTH_CONNECTION_STRING $TLS_CONNECTION_STRING PING)
+          
+          echo "The answer to PING is: $PONG"
+          echo "The ip is: $ip"
+          
+          if [[ -n $ip && $PONG == "PONG" ]];then
+            break
+          fi
+
+        done
+
+        redis-cli $AUTH_CONNECTION_STRING $TLS_CONNECTION_STRING CLUSTER MEET $ip $NODE_PORT
+        echo "Found $discrepancy IP discrepancy in line: $line"
+
+      fi
+
+    done < "$DATA_DIR/nodes.conf"
+
+    if [[ $discrepancy -eq 0 ]];then
+      echo "Did not find IP discrepancies between nodes."
+    fi
+
+  fi
+  return 0
+}
+
+ensure_replica_connects_to_the_right_master_ip(){
+  # This fixes an issue where a replica connects to the wrong ip of its master
+  # the node does not update the ip of its master and gets stuck trying to connect to an incorrect ip.
+  # To fix this we check for each slave if the master ip present (shown) using the "INFO REPLICATION"
+  # is also found in the /data/nodes.conf or in the "CLUSTER NODES" output and if it is not
+  # we update the new master using the CLUSTER REPLICATE command.
+  echo "Making sure slave is connected to master using right ip."
+  info=$(redis-cli $AUTH_CONNECTION_STRING $TLS_CONNECTION_STRING info replication)
+  if [[ "$info" =~ role:slave ]];then
+    master_ip=$(echo "$info" | grep master_host | cut -d':' -f2| tr -d '\r' )
+
+    ans=$(grep "$master_ip" "$DATA_DIR/nodes.conf")
+    if [[ -z $ans ]];then
+      echo "This instance is connected to its master using the wrong ip."
+      myself=$(grep 'myself' "$DATA_DIR/nodes.conf")
+      master_id=$(echo "$myself" | awk '{print $4}')
+      redis-cli $AUTH_CONNECTION_STRING $TLS_CONNECTION_STRING CLUSTER REPLICATE $master_id
+    fi
+
+  fi
+    
+}
+
+update_ips_in_nodes_conf(){
+  # Replace old ip with new one (external ip)
+  # This fixes the issue where when a node restarts it does not update its own ip
+  # this is fixed by getting the new public ip using the command "getent hosts $NODE_HOST" (NODE_HOST
+  # contains the domain name of the current node) and updating the nodes.conf file with the new ip before starting the redis server.
+  if [[ -f "$DATA_DIR/nodes.conf" && -s "$DATA_DIR/nodes.conf" ]];then
```

Comment:
We might have a problem for internal deployments, which are scheduled in private node pools. In that case, the correct IP address is the internal one, not the external.